pkg/queue/bounded_queue.go

```diff
@@ -79,11 +89,23 @@ func (q *BoundedQueue) Produce(item interface{}) bool {
 		q.onDroppedItem(item)
 		return false
 	}
+
+	// we might have two concurrent backing queues at the moment
+	// their combined size is stored in q.size, and their combined capacity
+	// should match the capacity of the new queue
+	if q.Size() >= q.Capacity() && q.Capacity() > 0 {
+		// current consumers of the queue (like tests) expect the queue capacity = 0 to work
+		// so, we don't drop items when the capacity is 0
+		q.onDroppedItem(item)
+		return false
+	}
```

Comment:
I don't understand this part. A second queue should never be the concern of Produce(), since it only operates on the _current_ queue. And why do we need capacity=0 for tests?

Reply:
> A second queue should never be the concern of Produce()

Correct, the comment is there only to make it explicit that there's no need to check two queues, as the state from both is handled at `q.Size()`. I can remove the comment if it's confusing.

> And why do we need capacity=0 for tests?

When no `Option.QueueSize()` is set and used in the span processor, the default queue size is 0 in tests like `cmd/collector/app/span_processor_test.go`. I changed the test to explicitly set a queue size, and removed the conditional here, as @pavolloffay had similar concerns.